File: BackTracking/Java/LC 90 - SUBSETS II.java
// LEET CODE 90 - SUBSETS II
// QUESTION LINK - https://leetcode.com/problems/subsets-ii/

import java.util.*;

class Solution {
    public void handleDup(int[] nums, List<List<Integer>> res, int start, List<Integer> temp) {
        res.add(new ArrayList<>(temp));

        for (int i = start; i < nums.length; i++) {
            if (i > start && nums[i] == nums[i - 1]) continue;
            temp.add(nums[i]);
            handleDup(nums, res, i + 1, temp);
            temp.remove(temp.size() - 1);
        }
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        handleDup(nums, res, 0, new ArrayList<>());
        return res;
    }
}

File: BackTracking/Python/LC 90 - SUBSETS II.py
# LEET CODE 90 - SUBSETS II
# QUESTION LINK - https://leetcode.com/problems/subsets-ii/

class Solution:
    def handleDup(self, nums, res, start, temp):
        res.append(temp[:])

        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i - 1]:
                continue
            temp.append(nums[i])
            self.handleDup(nums, res, i + 1, temp)
            temp.pop()

    def subsetsWithDup(self, nums):
        nums.sort()
        res = []
        self.handleDup(nums, res, 0, [])
        return res
